pkg/snmp/traps/format.go

```diff
@@ -0,0 +1,157 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://wwt.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package traps
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/soniah/gosnmp"
+)
+
+const (
+	sysUpTimeOID         = "1.3.6.1.2.1.1.3"
+	sysUpTimeInstanceOID = "1.3.6.1.2.1.1.3.0"
+	snmpTrapOID          = "1.3.6.1.6.3.1.1.4.1"
+	snmpTrapInstanceOID  = "1.3.6.1.6.3.1.1.4.1.0"
+)
+
+// NOTE: This module is used by the traps logs input.
+
+// FormatJSON converts an SNMP trap packet to a JSON-serializable object.
+func FormatJSON(p *SnmpPacket) (map[string]interface{}, error) {
+	return formatTrapPDUs(p.Content.Variables)
+}
+
+// GetTags returns a list of tags associated to an SNMP trap packet.
+func GetTags(p *SnmpPacket) []string {
+	tags := []string{
+		fmt.Sprintf("community:%s", p.Content.Community),
+		fmt.Sprintf("device_ip:%s", p.Addr.IP.String()),
+		fmt.Sprintf("device_port:%d", p.Addr.Port),
+	}
+	if version := formatVersion(p); version != "" {
+		tags = append(tags, fmt.Sprintf("snmp_version:%s", version))
+	}
+	return tags
+}
+
+func formatVersion(p *SnmpPacket) string {
+	switch p.Content.Version {
+	case gosnmp.Version2c:
+		return "2"
+	default:
+		return "" // Unsupported.
+	}
+}
+
+func formatTrapPDUs(vars []gosnmp.SnmpPDU) (map[string]interface{}, error) {
+	/*
+		An SNMPv2 trap packet consists in the following variables (PDUs):
+		{sysUpTime.0, snmpTrapOID.0, additionalDataVariables...}
+		See: https://tools.ietf.org/html/rfc3416#section-4.2.6
+	*/
+	if len(vars) < 2 {
+		return nil, fmt.Errorf("expected at least 2 variables, got %d", len(vars))
+	}
+
+	data := make(map[string]interface{})
+
+	uptime, err := parseSysUpTime(vars[0])
+	if err != nil {
+		return nil, err
+	}
+	data["uptime"] = uptime
+
+	trapOID, err := parseSnmpTrapOID(vars[1])
+	if err != nil {
+		return nil, err
+	}
+	data["oid"] = trapOID
+
+	data["variables"] = parseVariables(vars[2:])
+
+	return data, nil
+}
+
+func normalizeOID(value string) string {
+	return strings.TrimLeft(value, ".")
+}
+
+func parseSysUpTime(v gosnmp.SnmpPDU) (uint32, error) {
+	name := normalizeOID(v.Name)
+	if name != sysUpTimeOID && name != sysUpTimeInstanceOID {
```

Comment:
Is `sysUpTimeOID` a real use case without `.0` ? The RFC seems to indicate that the OID is an instance OID with `.0`.

Same comment for `snmpTrapOID`.

Reply:
I think given the mess that is SNMP, it's possible that some devices send it without the `.0`? I'm just being overly cautious here. 😅

Reply:
> I think given the mess that is SNMP, it's possible that some devices send it without the .0?

I can't say for sure, but probably unlikely, if it happen, it's a bug on the device. 

IMHO it's better to start with only support the instance OID (with `.0`) and add fallback if there are good reasons to support it later.

Also, we assume here, the OID is an instance OID with `.0`: https://github.com/DataDog/integrations-core/blob/e8eb6484a7aea40f5919929e02608cbe4babaacf/snmp/datadog_checks/snmp/config.py#L299

Reply:
Okay, so it looks like there's actually no such thing as `snmpTrapOIDInstance` - it's really just `snmpTrapOID` without a `.0`. But you're right, it should be `sysUptimeInstance` and never `sysUptime`. Fixed!

Reply:
Okay, so it looks like there's actually no such thing as `snmpTrapOIDInstance` - it's really just `snmpTrapOID` without a `.0`. But you're right, it should be `sysUptimeInstance` and never `sysUptime`. Fixed!